File: 2471. Minimum Number of Operations to Sort a Binary Tree by Level/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def min_swaps(self,nums:List[int]) -> int:
        idx_map = {}
        for idx,val in enumerate(nums):
            idx_map[val] = idx
        nums.sort()
        seen = set()
        swaps1 = 0
        for vals in idx_map:
            if vals != nums[idx_map[vals]]:
                x = vals
                cycle = 0
                while idx_map[x] not in seen:    
                    new_idx = idx_map[x]
                    seen.add(idx_map[x])
                    x = nums[new_idx] 
                    cycle += 1
                swaps1 += cycle - 1 if cycle > 1 else 0
         
        return swaps1    

    def minimumOperations(self, root: Optional[TreeNode]) -> int:
        queue = deque([root])
        swaps = 0
        while queue:
            n = len(queue)
            temp = []
            for i in range(n):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                    temp.append(node.left.val)
                if node.right:
                    queue.append(node.right)
                    temp.append(node.right.val)    
            swaps += self.min_swaps(temp)  
        return swaps          
    

        

File: 2471. Minimum Number of Operations to Sort a Binary Tree by Level/solution.rs
use std::collections::{HashMap, HashSet, VecDeque};
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn findswaps(arr: &mut Vec<i32>) -> i32 {
        let mut idx_map: HashMap<i32, usize> = HashMap::new();
        for (idx, &val) in arr.iter().enumerate() {
            idx_map.insert(val, idx);
        }

        arr.sort();

        let mut seen: HashSet<usize> = HashSet::new();
        let mut swaps = 0;

        for &val in idx_map.keys() {
            let mut cycles = 0;
            let mut x = val;

            while let Some(&idx_new) = idx_map.get(&x) {
                if seen.contains(&idx_new) {
                    break;
                }
                seen.insert(idx_new);
                x = arr[idx_new];
                cycles += 1;
            }

            if cycles > 1 {
                swaps += cycles - 1;
            }
        }

        swaps
    }

    pub fn minimum_operations(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut queue: VecDeque<Option<Rc<RefCell<TreeNode>>>> = VecDeque::new();
        queue.push_back(root.clone());
        let mut total_operations = 0;

        while !queue.is_empty() {
            let n = queue.len();
            let mut temp = vec![];

            for _ in 0..n {
                if let Some(Some(node)) = queue.pop_front() {
                    let node = node.borrow();

                    temp.push(node.val);

                    if let Some(left) = node.left.clone() {
                        queue.push_back(Some(left));
                    }
                    if let Some(right) = node.right.clone() {
                        queue.push_back(Some(right));
                    }
                }
            }

            total_operations += Self::findswaps(&mut temp);
        }

        total_operations
    }
}
